tasking/pulp/tasking/storage.py

```diff
@@ -0,0 +1,123 @@
+    """
+    Creates a working directory inside the working_directory as specified in /etc/pulp/server.yaml
+    default path for working_directory is /var/cache/pulp
+    :param worker_name:     Name of worker that uses the working directory created
+    :type  name:            basestring
+    :return Path to the working directory for a worker
+    :rtype  str
+    """
+    working_dir_root = _working_dir_root(worker_name)
+    os.mkdir(working_dir_root)
+
+
+def delete_worker_working_directory(worker_name):
+    """
+    Deletes a working directory inside the working_directory as specified in /etc/pulp/server.yaml
+    default path for working_directory is /var/cache/pulp
+    :param worker_name:     Name of worker that uses the working directory being deleted
+    :type  name:            basestring
+    """
+    working_dir_root = _working_dir_root(worker_name)
+    if os.path.exists(working_dir_root):
+        _rmtree_fix_permissions(working_dir_root)
+
+
+def _working_directory_path():
+    """
+    Gets the path for a task's working directory inside a workers working directory
+
+    @return: full path on disk to the working directory for current task
+    @rtype:  str
+    """
+    current_task = task.current
+    if (current_task and current_task.request and current_task.request.id and
+            current_task.request.hostname):
+        worker_name = current_task.request.hostname
+        task_id = current_task.request.id
+        worker_dir_root = _working_dir_root(worker_name)
+        working_dir_root = os.path.join(worker_dir_root, task_id)
+    else:
+        return None
+
+    return working_dir_root
+
+
+def delete_working_directory():
+    """
+    Deletes working directory for a particular task.
+
+    @return: None
+    @rtype:  None
+    """
+    working_dir_root = _working_directory_path()
+
+    if working_dir_root and os.path.exists(working_dir_root):
+        _rmtree_fix_permissions(working_dir_root)
+
+
+def get_working_directory():
+    """
+    Creates a working directory with task id as name. This directory resides inside a particular
+    worker's working directory. The 'working_directory' setting in ``server`` section of
+    /etc/pulp/server.yml defines the local path to the root of working directories. The directory
+    is created only once per task. The path to the existing working directory is returned for all
+    subsequent calls for that task.
+
+    @return: full path on disk to the working directory for current task
+    @rtype:  str
+    """
+    working_dir_root = _working_directory_path()
+
+    if working_dir_root:
+        if os.path.exists(working_dir_root):
+            return working_dir_root
+        os.mkdir(working_dir_root)
+        return working_dir_root
+    else:
+        # If path is None, this method is called outside of an asynchronous task
+        raise RuntimeError("Working Directory requested outside of asynchronous task.")
+
+
+def _rmtree_fix_permissions(directory_path):
```

Comment:
I think we should keep this behavior when removing working directories, but the comment could be improved to give better context.

I believe the original problem this addressed was that when extracting the contents of a puppet module, the permissions on files could be unexpected. We've seen a surprising number of such modules. Cleanup of the working directory was not able to remove some files without first changing their permissions.

That particular plugin now makes efforts to fix those permissions so they are reasonable, but it's a best-effort. The plugin does not guarantee that all files it leaves behind will have reasonable permissions.

Any other plugin that engages in extraction of files from an archive could face the same issue. Having the platform be defensive in this way is reasonable I think.